Fixes #91.  Make IO resources not autoclose.

enebo opened the pull request:
A later JRuby release may allow these autoclose=false statements to be
removed (although @handle may desire to keep it's due to to_io handing
that IO out to external consumers).  The basic problem is currently during
shutdown JRuby Threads may not have finished executing before finalizers
start firing.  JRuby is looking at sending a more explicit termination
request similar to MRI to threads on shutdown, which may obviate the need
for this.  That change is too risky in the near term so we are requesting
this PR for the time being.

matthewd commented:
I'm not sure how well this fixes #91... at a glance, I think the change as written solves the immediate issue by turning it into a handle leak (the IOs won't be closed even after the Notifier is GCed), and AIUI introducing a manual finalizer to solve that would just put us back in the same situation, where during shutdown JRuby will freely invoke a finalizer despite another still-running thread holding an active reference to the object.

Am I missing something that makes this safer than it appears? Alternatively, if we're going to have to introduce a leak, I think we should constrain it to JRuby.

enebo commented:
@matthewd This is a long comment but there is a lot to unpack.  Apologies in advance :)

Firstly, I agree with you that we should not change behavior for any MRI uses but I do have a section below called determinism where I talk a bit more about an issue I see.  So I will add JRuby guards on those lines to reduce amount of possible surprises for rb-inotify.  As I mentioned in #91 we are working on a patch which should eventually remove the needs of autoclose = false by trying to more explicitly shut down the threads before finalization starts (we did not land for 9.2.6.0 because it is a change where we need bake time).

**On why the PR works**

For JRuby 9.2.6.0 we removed a call which closes all IOs during shutdown directly.  This was JRuby at shutdown of the Ruby runtime explicitly closing IOs which should have been explicitly closed by the original IO user or ended up hitting Java finalization (which happens after our finalization).   We removed this code so close is not specifically an issue from 9.2.6.0 onwards.  I could talk about why this code should have never existed but suffice it to say IOError closed stream part of the problem went away.

This leads to the next problem that for select() we have an internal selector pool and that pool does close all the open channels during Ruby shutdown.  This PR is what works around this. Otherwise in same places we saw IOError: stream closed we start getting selector closed errors.  By setting auto_close = false @handler and pipe then do not get closed so select() continues to work until Java really shuts down and fires its finalization.  Java finalization cannot occur on an object still in use so it won't be an issue by that point.

**Manual finalizer**

I think a manual finalizer should work because those are fired before any other of our internal finalization and calling .stop uses synchronization and close these resources before it continues to internal finalization.

**Determinism**

(apologies if this is old hat)

GC semantically is non-deterministic in when it fires (real time systerms with GC do guarantee determinism but Ruby is not real time).  In MRI, for better or worse, it tends to fire so much that it appears to be deterministic.  In the future, if MRI changes the implementation the assumption that finalization will occur in a timely matter may stop working.  More on that a little below.

For JRuby, we already have the leak in an implicit let GC finalize scenario.   JVM GCs tend to be as lazy as possible for firing off collections which gives much better GC throughput but that also means when an object is noticed to actually be dead may take a long time.  Objects which use java finalization take even longer to actually finalize than ordinary objects.  The time it takes to GC is mostly a function of memory allocated and the more memory the less collections which means less time spent on memory management.  In this sense, the tradeoff is that throughput is valued over time to notice an object is dead. finalization is probably the only negative aspect of this tradeoff...

If Ruby programs idiomatic usage of resources is to assume the GC will handle finalization then I fear MRI will decide they can never improve their GC throughput by delaying collections.   This is especially important today as they are slowly making their GC generational.  This is not really rb-inotify's problem since you have a nice explicit close mechanism but I wonder if there is something we can do to discourage or educate from using implicit finalization?

If people were explicitly using close while using rb-inotify when finished or registering an at_exit block then JRuby would always work as expected and this implicit pattern would potentially be removed as roadblock from improving MRIs GC.  For this project this may only be some documentation.  Possibly notifying a few users they could be closing things more explicitly?

Apologies for being so long-winded...This last section is not strictly related to the PR but when you mentioned introducing a leak I felt I wanted to explain that it is already leaky.

afromankenobi commented:
Hi! Any notice on this? we are facing the exact same problem and this solution makes a lot of sense.

ioquatix commented:
I will review it.

ioquatix commented:
I am merging this I will review it now and hopefully we can release.
